include/swift/SIL/SILInstruction.h

```diff
@@ -1446,6 +1446,55 @@ class AllocRefInst final
   }
 };
 
+/// ReallocRefInst - This represents the reallocation primitive of a reference
+/// type. If the reallocation was successfully the reference is returned with
+/// the altered size. If the reallocation failed the reference is returned at
+/// a new address in an allocation with the altered size and the contents are
+/// bit-wise copied.
+/// This is only useful for tail-allocated arrays and the new size must always
+/// be larger than the original size.
+class ReallocRefInst final
+    : public InstructionBaseWithTrailingOperands<
+                                               SILInstructionKind::ReallocRefInst,
+                                               ReallocRefInst,
+                                               AllocRefInstBase, SILType> {
+  friend AllocRefInstBase;
+  friend SILBuilder;
+
+  ReallocRefInst(SILDebugLocation DebugLoc, SILFunction &F, SILType ObjectType,
+                 SILValue ExistingRef, ArrayRef<SILType> ElementTypes,
+                 ArrayRef<SILValue> AllOperands)
+      : InstructionBaseWithTrailingOperands(AllOperands, DebugLoc, ObjectType,
+                                            false, false, ElementTypes) {
+    assert(AllOperands.size() >= ElementTypes.size());
+    std::uninitialized_copy(ElementTypes.begin(), ElementTypes.end(),
+                            getTrailingObjects<SILType>());
+  }
+
+  static ReallocRefInst *create(SILDebugLocation DebugLoc, SILFunction &F,
+                                SILType ObjectType, SILValue ExistingRef,
+                                ArrayRef<SILType> ElementTypes,
+                                ArrayRef<SILValue> ElementCountOperands,
+                                SILOpenedArchetypesState &OpenedArchetypes);
+
+public:
+  ArrayRef<Operand> getTypeDependentOperands() const {
+    return getAllOperands().slice(getNumTailTypes());
+  }
+
+  MutableArrayRef<Operand> getTypeDependentOperands() {
+    return getAllOperands().slice(getNumTailTypes());
+  }
+
+  SILValue getReallocPointer() {
```

Comment:
Btw, I just made up these names, feel free to use your intuition for better names. The names here that I made up I feel fail the DRY test since realloc is in the class name. Also, generally the way that this sort of thing is implemented is that you define an enum that defines the index and then use that with getOperand(unsigned) to get the right value. See StoreInst which follows that pattern.